### ISSUE TITLE
Warn about missing Flutter SDK constraint and missing upper bound on flutter SDK constraint

### ISSUE THREAD

**Issue by jonasfj** [dart-lang/pub#2659]

**Comment by sigurdm** [comment]
I don't think we want to do the upper bound - flutter no longer interprets the upper bound.

**Comment by sigurdm** [comment]
Could this be implemented in the analyzer or as a hint somehow?

That would give the warning earlier.

Question is, how to detect that we rely on the flutter sdk?

**Comment by pq** [comment]
This could certainly and probably ideally would be done in the analyzer.

As for detecting flutter-ness, I wonder... I can think of some heuristics but I bet there's some prior art here.  Maybe @goderbauer knows?

**Comment by sigurdm** [comment]
Yeah - in pub we could say that if any package in the resolution comes from the flutter sdk it is a flutter package - but I'm not sure that information is available to the analyzer.

Maybe it can be detected that `dart:ui` is in the import path somewhere...

**Comment by pq** [comment]
> in pub we could say that if any package in the resolution comes from the flutter sdk

I'm not sure I understand this but it feels promising...  Is this information in the `package_config.json`?

Popping up, I realize I'm not exactly sure what we want to check here. 😀

Are we looking for a _missing_  bit like this in a pubspec?


```yaml
dependencies:
  flutter:
    sdk: flutter
```

**Comment by sigurdm** [comment]
> Popping up, I realize I'm not exactly sure what we want to check here. 😀

:) I guess we should paint the details a bit better:

To indicate that a package works with a given flutter release a flutter package should have a flutter sdk constraint like:
```
environment:
  flutter: '>3.1.0'
```

That is the missing bit.

However a package can leave out that constraint, and still depend on flutter:
* via dependencies from the flutter sdk
  ```
  dependencies:
    flutter:
      sdk: flutter # <-- This line only works with a flutter sdk
  ```
* via other dependencies that depend on flutter
  ```
  dependencies:
    camera: # <-- the camera package depends on flutter
  ```
* Via import of `dart:ui` in a dart file.
  ```
   import 'dart:ui'; // <-- the containing class depends on flutter.
  ```

Ideally we would like the diagnostic to cover all these cases, and suggest adding an sdk constraint on Flutter.

**Comment by pq** [comment]
Ah!  This is **very** useful.  Thanks for the clarification!

> Ideally we would like the diagnostic to cover all these cases, and suggest adding an sdk constraint on Flutter.

I agree.
